src/Gui/ComputationDialog.cpp

```diff
@@ -0,0 +1,162 @@
+}
+
+void ComputationDialog::abort() {
+    aborted.store(true);
+    reject();
+}
+
+void ComputationDialog::run(std::function<void()> func) {
+    std::atomic<bool> computationDone(false);
+    std::mutex mutex;
+    std::condition_variable cv;
+    std::exception_ptr threadException;
+
+    Base::ProgressIndicator::setInstance(this);
+
+    QMessageBox forceAbortBox(
+        QMessageBox::Warning,
+        QObject::tr("Operation not responding"),
+        QObject::tr("Aborting the operation is taking longer than expected.\nDo you want to forcibly cancel the thread? This will probably crash FreeCAD."),
+        QMessageBox::Yes | QMessageBox::No,
+        Gui::MainWindow::getInstance());
+    forceAbortBox.setDefaultButton(QMessageBox::No);
+
+    // Start computation thread
+    std::thread computeThread([&]() {
+        #if defined(__linux__) || defined(__FreeBSD__)
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, nullptr);
+        pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, nullptr);
+        #endif
+
+        try {
+            func();
+        } catch (...) {
+            threadException = std::current_exception();
+        }
+
+        computationDone.store(true);
+
+        if (isVisible()) {
+            QMetaObject::invokeMethod(this, "accept", Qt::QueuedConnection);
+        }
+        if (forceAbortBox.isVisible()) {
+            QMetaObject::invokeMethod(&forceAbortBox, "accept", Qt::QueuedConnection);
+        }
+
+        cv.notify_one();
+    });
+
+    {
+        // Wait for a brief moment to see if computation completes quickly
+        std::unique_lock<std::mutex> lock(mutex);
+        if (!cv.wait_for(lock, std::chrono::seconds(1),
+            [&]{ return computationDone.load(); }))  // Atomic load
+        {
+            // Computation didn't finish quickly, show dialog
+            exec();
+        }
+    }
+
+    while (!computationDone.load()) {
+        // Wait for up to 3 seconds for the thread to finish
+        std::unique_lock<std::mutex> lock(mutex);
+        if (!cv.wait_for(lock, std::chrono::seconds(3),
+            [&]{ return computationDone.load(); }))
+        {
+            if (forceAbortBox.exec() == QMessageBox::Yes) {
+                // check computationDone again just in case the thread completed while the dialog was open
+                if (!computationDone.load()) {
+                    Base::Console().Error("Force aborting computation thread\n");
+
+                    // TODO: save the backup document now!
+
+                    #if defined(__linux__) || defined(__FreeBSD__)
+                    pthread_cancel(computeThread.native_handle());
+                    #elif defined(_WIN32)
+                    TerminateThread(computeThread.native_handle(), 1);
+                    CloseHandle(computeThread.native_handle());
+                    #elif defined(__APPLE__)
+                    pthread_kill(computeThread.native_handle(), SIGTERM);
+                    #endif
```

Comment:
The posix thread support on macOS is compatible enough to use the same codepath as linux and freebsd. We can simplify this to the following (tested on macOS):
```suggestion
#if defined(FC_OS_WIN32)
                    TerminateThread(computeThread.native_handle(), 1);
                    CloseHandle(computeThread.native_handle());
#else
                    pthread_cancel(computeThread.native_handle());
#endif
```

Reply:
Thanks, done.